### PR TITLE
[Bugfix:CourseMaterials] Course materials enumerated

### LINKED ISSUE #8645: Course materials enumerated on every request
**Describe the bug**
It appears as though course materials are enumerated on every request.
While I was trying to discover where certain files were uploaded, I ran the following:
`inotifywait -r -m /var/local/submitty/courses`
I noticed that >600 filesystem accesses were performed during a file upload.
After some investigation, I found that even on requests to the discussion forum, course materials were being enumerated.
This may not be a huge performance hit since it's not reading all the files, but maybe something Submitty shouldn't be doing.
Logs:
```
/var/local/submitty/courses/f22/sample/config/ OPEN config.json
/var/local/submitty/courses/f22/sample/config/ ACCESS config.json
/var/local/submitty/courses/f22/sample/config/ CLOSE_NOWRITE,CLOSE config.json
/var/local/submitty/courses/f22/sample/uploads/ OPEN,ISDIR course_materials
/var/local/submitty/courses/f22/sample/uploads/course_materials/ OPEN,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/ ACCESS,ISDIR course_materials
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/ ACCESS,ISDIR course_materials
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ OPEN,ISDIR outer
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ OPEN,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR outer
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR outer
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ OPEN,ISDIR inner
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/inner/ OPEN,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ ACCESS,ISDIR inner
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/inner/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ ACCESS,ISDIR inner
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/inner/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ ACCESS,ISDIR inner
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/inner/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ CLOSE_NOWRITE,CLOSE,ISDIR inner
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/inner/ CLOSE_NOWRITE,CLOSE,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR outer
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ CLOSE_NOWRITE,CLOSE,ISDIR outer
/var/local/submitty/courses/f22/sample/uploads/course_materials/outer/ CLOSE_NOWRITE,CLOSE,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ OPEN,ISDIR words
/var/local/submitty/courses/f22/sample/uploads/course_materials/words/ OPEN,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR words
/var/local/submitty/courses/f22/sample/uploads/course_materials/words/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR words
/var/local/submitty/courses/f22/sample/uploads/course_materials/words/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR words
/var/local/submitty/courses/f22/sample/uploads/course_materials/words/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/course_materials/ CLOSE_NOWRITE,CLOSE,ISDIR words
/var/local/submitty/courses/f22/sample/uploads/course_materials/words/ CLOSE_NOWRITE,CLOSE,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/ ACCESS,ISDIR course_materials
/var/local/submitty/courses/f22/sample/uploads/course_materials/ ACCESS,ISDIR 
/var/local/submitty/courses/f22/sample/uploads/ CLOSE_NOWRITE,CLOSE,ISDIR course_materials
/var/local/submitty/courses/f22/sample/uploads/course_materials/ CLOSE_NOWRITE,CLOSE,ISDIR 
```

Additionally, when a gradeable is visited, EVERY assignment's `config/build` json file is being read, not just for the assignment at hand.

### PR DESCRIPTION
**Pull request by 04vmatsibekker**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Closes #8645.
When the site is trying to figure out if the course_materials dir folder is empty, it goes through all the files to check.

### What is the new behavior?
Now we only check to see if there is at least 1 file in course_materials dir.

### CODE REVIEW

File: site/app/libraries/FileUtils.php [hunk 1]
```diff
@@ -174,6 +174,24 @@ public static function emptyDir($dir) {
...
+            while (false !== ($file = readdir($handle))) {
+                if ($file != '.' && $file != '..') {
+                    closedir($handle);
+                    return false;
+                }
+            }
+            closedir($handle);
+        }
+        return true;
+    }
```
Comment (williamjallen) [inline site/app/libraries/FileUtils.php:193]: Would it not be cleaner to use `return count(scandir(...)) > 2` for this?

File: site/app/libraries/FileUtils.php [hunk 2]
```diff
@@ -174,6 +174,24 @@ public static function emptyDir($dir) {
         }
     }
 
+    /**
+     * checks to see if directory is empty, if so return true
+     *
+     * @param string $dir
+     */
+    public static function isEmptyDir($dir) {
```
Comment (williamjallen) [inline site/app/libraries/FileUtils.php:182]: Please add native PHP type declarations.

```suggestion
     */
    public static function isEmptyDir(string $dir): bool {
```

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/libraries/FileUtils.php (hunk 1 above)
Comment (04vmatsibekker) [reply site/app/libraries/FileUtils.php:193]: I saw that scandir did look at all of the files, perhaps I am wrong. But that is why I chose to write a function to check without looking at all of the files if the directory is empty.

Comment (NiaHeermance) [reply site/app/libraries/FileUtils.php]: Viane is correct. scandir would defeat the point of this PR, which is to make it so we _don't_ have to look at every file to determine if the course_materials directory is empty.

**Review by cjreed121 — COMMENTED** [review]
This definitely seems to fix the issue. Because the database and directory should always be synced, we could detect if there exists any course materials using the database instead. Although, it's tough to say if a database call or file system access will be more expensive. While database call requires networking which may seem slower, a slow filesystem could potentially take longer. I think a little bit of design discussion is needed.

**Pull request merged**